unikernel.ml

```diff
@@ -128,66 +142,128 @@ struct
+                  ~with_error:true ~msg:"Session cookie is no longer valid."
+                  reqd ())
+              else
+                Middleware.apply_middleware
+                  (middlewares ~form_csrf user)
+                  (fun reqd ->
+                    let cookie =
+                      { cookie with user_agent = Middleware.user_agent reqd }
+                    in
+                    let cookies =
+                      List.map
+                        (fun (cookie' : User_model.cookie) ->
+                          if String.equal cookie_value cookie'.value then cookie
+                          else cookie')
+                        user.cookies
+                    in
+                    let updated_user =
+                      User_model.update_user user ~cookies ()
+                    in
+                    Store.update_user store updated_user >>= function
+                    | Ok () -> f ~json_dict user
+                    | Error (`Msg err) ->
+                        Logs.err (fun m -> m "Error with storage: %s" err);
+                        Middleware.http_response reqd ~title:"Error"
+                          ~data:(`String err) `Not_found)
+                  reqd)
+    in
+    match (Middleware.api_authentication reqd, api_meth) with
+    | Some token_value, true -> (
+        extract_json_body reqd >>= function
+        | Ok json_dict -> (
+            match Store.find_by_api_token store token_value with
+            | Some (user, token) ->
+                if User_model.is_valid_token token now then (
+                  let token =
+                    { token with usage_count = token.usage_count + 1 }
                   in
-                  let cookies =
+                  let tokens =
                     List.map
-                      (fun (cookie' : User_model.cookie) ->
-                        if String.equal cookie_value cookie'.value then cookie
-                        else cookie')
-                      user.cookies
+                      (fun (token' : User_model.token) ->
+                        if String.equal token_value token'.value then token
+                        else token')
+                      user.tokens
                   in
-                  let updated_user = User_model.update_user user ~cookies () in
+                  let updated_user = User_model.update_user user ~tokens () in
                   Store.update_user store updated_user >>= function
-                  | Ok () -> f user
+                  | Ok () -> f ~json_dict user
                   | Error (`Msg err) ->
                       Logs.err (fun m -> m "Error with storage: %s" err);
-                      Middleware.http_response reqd ~title:"Error" ~data:err
-                        `Not_found)
-                reqd)
+                      Middleware.http_response reqd ~title:"Error"
+                        ~data:(`String err) `Not_found)
```

Comment:
can this please be moved to a function in storage (or parts in user_model)? I don't think in `authenticate` we should think about "how to increment a token counter"...